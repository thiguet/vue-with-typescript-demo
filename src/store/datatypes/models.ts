interface Unique {
    id?: string;
}

export enum Measures {
    UN = 'UN',
    L = 'L',
    KG = 'KG',
    G = 'G',
}

export interface Product extends Unique {
    name: string;
    measure: Measures;
    qtd: number;
    minQtd: number;
    image?: string;
}

export interface User extends Unique {
    name: string;
    email: string;
}

export interface ReportItem {
    text: string;
    value: number;
}

export type ReportData = ReportItem[];

export interface QuantityReportData {
    productName: string;
    productQuantity: number;
}

export type QuantityReport = QuantityReportData[];
export interface MeasuresReportData {
    measureName: string;
    measureQuantity: number;
}

export type MeasuresReport = MeasuresReportData[];

export enum VuexAppModules {
    products = 'products',
    login = 'login',
    alert = 'alert',
    reports = 'reports',
}
