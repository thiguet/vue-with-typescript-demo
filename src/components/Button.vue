<template>
    <div class="wrapper-btn">
        <button class="btn"
                :name="name"
                @click="onclick">{{ label }}</button>
    </div>
</template>

<script lang="ts">
import Vue from 'vue';

import Component from 'vue-class-component';
import { Prop } from 'vue-property-decorator';

@Component
export default class Button extends Vue {
    @Prop({ required: true }) readonly label!: string;

    @Prop({ required: true }) readonly onclick!: Function;
}
</script>
