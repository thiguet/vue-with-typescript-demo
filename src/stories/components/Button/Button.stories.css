@import url('../../app.css');

.btn {
    background-color: rgb(204, 52, 49) !important;
}

.icon {
    width: 100px;
    height: 100px;
}
