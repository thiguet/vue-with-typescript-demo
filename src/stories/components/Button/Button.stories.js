import Button from '@/components/Button.vue';
import './Button.stories.css';
import imageFile from './logo.gif';

export default {
    title: 'Example/Components/Button',
    component: Button,
};

const Template = (_args, { argTypes }) => ({
    props: Object.keys(argTypes),
    components: { Button },
    template: '<Button @onclick="onclick" v-bind="$props" />',
});

export const Default = Template.bind({});
Default.args = {
    id: 'some-id',
    name: 'some-name',
    label: 'Button',
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    onclick: () => {},
};

export const WithIcon = Template.bind({});
WithIcon.args = {
    id: 'some-id2',
    name: 'some-name2',
    label: 'Button With Icon',
    icon: imageFile,
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    onclick: () => {},
};
