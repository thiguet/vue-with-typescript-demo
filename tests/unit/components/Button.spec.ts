import { shallowMount } from '@vue/test-utils';
import Button from '@/components/Button.vue';
import faker from 'faker';
import { ButtonProps } from './models';

describe('Button', () => {
    let props: ButtonProps;

    const build = () => {
        const wrapper = shallowMount(Button, {
            propsData: { ...props },
        });

        return {
            wrapper,
            btnWrapper: () => wrapper.find('.wrapper-btn'),
            button: () => wrapper.find('.btn'),
            icon: () => wrapper.find('.icon'),
        };
    };

    beforeEach(() => {
        props = {
            id: faker.random.uuid(),
            name: faker.random.word(),
            label: faker.random.word(),
            icon: faker.image.imageUrl(),
            onclick: jest.fn(),
        };
    });

    it('renders component', () => {
        const fixedProps: ButtonProps = {
            id: 'someID',
            name: 'someName',
            icon: 'some broken icon',
            label: 'someLabel',
            onclick: jest.fn(),
        };
        props = fixedProps;

        const { wrapper } = build();

        expect(wrapper).toMatchSnapshot();
    });

    it('check if props have been passed', () => {
        const { wrapper } = build();

        expect({ ...props }).toEqual({ ...wrapper.vm.$props });
    });

    it('renders main child components', () => {
        const { btnWrapper, button, icon } = build();

        expect(btnWrapper().exists()).toBe(true);
        expect(button().exists()).toBe(true);
        expect(icon().exists()).toBe(true);
    });

    it('clicks method', async () => {
        const { button } = build();

        await button().trigger('click');

        expect(props.onclick).toHaveBeenCalled();
    });

    it('sets buttons name.', async () => {
        const { button } = build();

        expect(props.name).toBe(button().attributes().name);
    });

    it('must call evt.preventDefault fn', async () => {
        const { button } = build();

        const opts = {
            preventDefault: jest.fn(),
        };

        button().trigger('click', opts);

        expect(opts.preventDefault).toHaveBeenCalled();
    });

    it('must set icon to the img src', async () => {
        const { icon } = build();

        expect(icon().attributes('src')).toBe(props.icon);
    });
});
