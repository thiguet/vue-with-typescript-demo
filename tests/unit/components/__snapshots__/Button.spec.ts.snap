// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders component 1`] = `<div class="wrapper-btn"><button class="btn">someFixedLabel</button></div>`;
